### PR TITLE
Feature/748 validation layer

### PR DESCRIPTION
**Pull request by eskebab**
## Related Issue(s)
- #748 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

## Summary by CodeRabbit

- **New Features**
  - Enhanced validation across notification channels improves error feedback and ensures consistent delivery of email and SMS notifications.
  - Introduced new validators for email and SMS notification requests, recipient details, reminders, and notification order chains, enhancing validation robustness.
  - Added a new abstract base class for notification recipients, streamlining delivery channel configurations.
  - New test classes and methods provide comprehensive coverage for notification validation logic.
- **Refactor**
  - Reorganized notification models and namespaces for better structure and maintainability.
  - Simplified validator registration by adding assembly-wide registration with duplicate checks.
  - Removed the `NotificationOrderWithRemindersRequestValidator` to consolidate validation logic.

These updates improve notification reliability and maintainability while enhancing validation clarity and coverage.

### CODE REVIEW

File: src/Altinn.Notifications/Validators/DialogportenIdentifiersValidator.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    /// Class containing validation logic for the <see cref="DialogportenIdentifiersExt"/> model
+    /// </summary>
+    internal sealed class DialogportenIdentifiersValidator : AbstractValidator<DialogportenIdentifiersExt?>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DialogportenIdentifiersValidator"/> class.
+        /// </summary>
+        public DialogportenIdentifiersValidator()
+        {
+        }
```
[the remark this answers is not in the document]
Comment (eskebab) [reply src/Altinn.Notifications/Validators/DialogportenIdentifiersValidator.cs:17]: this solution is not relevant

[the remark this answers is not in the document]
Comment (eskebab) [reply src/Altinn.Notifications/Validators/DialogportenIdentifiersValidator.cs:17]: I have left it empty intentionally for now, but I might as well delete it, I suppose. what do you think? Is it better to delete it and then re-create it later, if needed in the future?

[the remark this answers is not in the document]
Comment (eskebab) [reply src/Altinn.Notifications/Validators/DialogportenIdentifiersValidator.cs:17]: thats a good idea, thanks. can you create a commit with that comment?

File: src/Altinn.Notifications/Validators/Recipient/RecipientBaseValidator.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+/// <summary>
+/// Represents validation logic for the recipient base model.
+/// </summary>
+public class RecipientBaseValidator : AbstractValidator<RecipientBaseExt?>
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RecipientBaseValidator"/> class.
+    /// </summary>
+    public RecipientBaseValidator()
```
Comment (Ahmed-Ghanam) [inline src/Altinn.Notifications/Validators/Recipient/RecipientBaseValidator.cs:19]: We need to determine whether we should implement a validation rule for the `resourceId` to ensure it is always a value that the API can process.

File: src/Altinn.Notifications/Validators/NotificationOrderBaseValidator.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    /// Validator for <see cref="NotificationOrderBaseExt"/>.
+    /// </summary>
+    internal sealed class NotificationOrderBaseValidator : AbstractValidator<NotificationOrderBaseExt>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NotificationOrderBaseValidator"/> class.
+        /// </summary>
+        public NotificationOrderBaseValidator()
+        {
+            RuleFor(option => option.RequestedSendTime)
```
Comment (Ahmed-Ghanam) [inline src/Altinn.Notifications/Validators/NotificationOrderBaseValidator.cs:17]: The `RequestedSendTime` must have specified a time zone. See `NotificationOrderRequestBaseRules`

File: src/Altinn.Notifications/Validators/NotificationOrderBaseValidator.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /// <summary>
+    /// Validator for <see cref="NotificationOrderBaseExt"/>.
+    /// </summary>
+    internal sealed class NotificationOrderBaseValidator : AbstractValidator<NotificationOrderBaseExt>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NotificationOrderBaseValidator"/> class.
+        /// </summary>
+        public NotificationOrderBaseValidator()
```
Comment (Ahmed-Ghanam) [inline src/Altinn.Notifications/Validators/NotificationOrderBaseValidator.cs:15]: We need to check that the `ConditionEndpoint` is a valid URL.

File: src/Altinn.Notifications/Program.cs [hunk 5]
```diff
@@ -288,10 +286,7 @@ async Task ConnectToKeyVaultAndSetApplicationInsights(ConfigurationManager confi
 void AddInputModelValidators(IServiceCollection services)
 {
     ValidatorOptions.Global.LanguageManager.Enabled = false;
-    services.AddSingleton<IValidator<EmailNotificationOrderRequestExt>, EmailNotificationOrderRequestValidator>();
-    services.AddSingleton<IValidator<SmsNotificationOrderRequestExt>, SmsNotificationOrderRequestValidator>();
-    services.AddSingleton<IValidator<NotificationOrderRequestExt>, NotificationOrderRequestValidator>();
-    services.AddSingleton<IValidator<NotificationOrderChainRequestExt>, NotificationOrderWithRemindersRequestValidator>();
+    services.AddValidatorsFromAssembly(typeof(Program).Assembly, includeInternalTypes: true);
```
Comment (Ahmed-Ghanam) [inline src/Altinn.Notifications/Program.cs:289]: To avoid conflicts, we must ensure that there is no more than one validator for each type. This will prevent issues that may arise if multiple validators exist for the same type.

**Review by Ahmed-Ghanam — CHANGES_REQUESTED** [review]

File: src/Altinn.Notifications/Validators/Email/EmailSendingOptionsValidator.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /// <summary>
+    /// Represents validation logic for the email sending options.
+    /// </summary>
+    internal sealed class EmailSendingOptionsValidator : AbstractValidator<EmailSendingOptionsExt?>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EmailSendingOptionsValidator"/> class.
+        /// </summary>
+        public EmailSendingOptionsValidator()
```
[the remark this answers is not in the document]
Comment (eskebab) [reply src/Altinn.Notifications/Validators/Email/EmailSendingOptionsValidator.cs:17]: since it's an enum with a default value, it is limited what we can do, as long as both values are permitted.
If a value is provided that isn't in the enum, the deserializer will throw an error before the validator is called.

File: src/Altinn.Notifications/Validators/Recipient/RecipientBaseValidator.cs (hunk 2 above)
Comment (eskebab) [reply src/Altinn.Notifications/Validators/Recipient/RecipientBaseValidator.cs:19]: that's a good idea

File: src/Altinn.Notifications/Program.cs (hunk 5 above)
Comment (eskebab) [reply src/Altinn.Notifications/Program.cs:289]: true that. I will try make a proper check for duplicates

File: src/Altinn.Notifications/Validators/Email/EmailSendingOptionsValidator.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ahmed-Ghanam) [reply src/Altinn.Notifications/Validators/Email/EmailSendingOptionsValidator.cs:17]: It would be good practice to add validation for the `ContentType` property to ensure that only valid enum values are accepted. Even though .NET will reject invalid enum values during deserialization, adding explicit validation provides more descriptive error messages to API consumers.

**Comment by Ahmed-Ghanam** [comment]
We need to increase the number of unit tests to meet the minimum test coverage requirement. CodeRabbit has provided several valuable comments that we should consider. I have approved the changes, allowing us to thoroughly test them in the test environments after implementing the necessary modifications and improvements.

**Review by Ahmed-Ghanam — APPROVED** [review]
~~Perhaps approving the changes was premature, as there is still some work that needs to be completed.~~

**Review by Ahmed-Ghanam — CHANGES_REQUESTED** [review]
Perhaps approving the changes was premature, as there is still some work that needs to be completed.

**Comment by Ahmed-Ghanam** [comment]
@eskebab, I noticed that CodeRabbit has provided several valuable suggestions regarding the validation rules implemented so far. Please review them here: https://github.com/Altinn/altinn-notifications/pull/796#pullrequestreview-2783415362

**Review by Ahmed-Ghanam — APPROVED** [review]
Good work :)
